File: app/src/main/java/com/example/TranslationEngine.kt
package com.example.simultrans

import com.google.ai.edge.litertlm.Backend
import com.google.ai.edge.litertlm.Content
import com.google.ai.edge.litertlm.Contents
import com.google.ai.edge.litertlm.ConversationConfig
import com.google.ai.edge.litertlm.Engine
import com.google.ai.edge.litertlm.EngineConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.io.File

/**
 * Envoltorio simple sobre el motor de LiteRT-LM para hacer traducciones
 * puntuales (sin memoria entre turnos) y consultas libres (con o sin
 * imágenes adjuntas) con el modelo Gemma 4 E2B.
 *
 * Cada llamada a [translate] o [ask] abre una conversación nueva, para que
 * el modelo no arrastre contexto de turnos anteriores.
 */
class TranslationEngine(private val modelFile: File) {
    private lateinit var engine: Engine
    private val mutex = Mutex()
    val isModelPresent: Boolean
        get() = modelFile.exists() && modelFile.length() > 0

    suspend fun initialize() = withContext(Dispatchers.IO) {
        val config = EngineConfig(
            modelPath = modelFile.absolutePath,
            // CPU es lo más compatible entre dispositivos. Si tu app solo
            // apunta a gama media/alta, puedes cambiar a Backend.GPU() para
            // más velocidad (requiere declarar libOpenCL.so en el manifest).
            backend = Backend.CPU(),
            // Necesario para que el modelo pueda procesar las imágenes que
            // se adjuntan desde la pestaña "Asistente IA" (fotos o páginas
            // de PDF rasterizadas). Gemma 4 E2B es multimodal de fábrica.
            visionBackend = Backend.CPU(),
            // Necesario para el reconocimiento de voz propio (transcribe):
            // en vez de depender del SpeechRecognizer del sistema (que en
            // algunos fabricantes, como Samsung, no respeta los paquetes
            // offline instalados), Gemma 4 E2B reconoce audio de forma
            // nativa sin salir de la app.
            audioBackend = Backend.CPU(),
        )
        engine = Engine(config)
        engine.initialize()
    }

    suspend fun translate(text: String, fromLangName: String, toLangName: String): String =
        withContext(Dispatchers.IO) {
            mutex.withLock {
                val systemInstruction = """
                    Eres un traductor profesional simultáneo.
                    Traduce SIEMPRE del $fromLangName al $toLangName.
                    Responde ÚNICAMENTE con la traducción, sin explicaciones,
                    sin comillas y sin repetir el texto original.
                """.trimIndent()
                val conversationConfig = ConversationConfig(
                    systemInstruction = Contents.of(systemInstruction),
                )
                engine.createConversation(conversationConfig).use { conversation ->
                    conversation.sendMessage(text).toString().trim()
                }
            }
        }

    /**
     * Consulta libre al modelo, con 0 o varias imágenes adjuntas (fotos
     * sueltas, o páginas de un PDF ya rasterizadas a imagen por quien
     * llama a esta función). No lleva instrucción de sistema fija: el
     * usuario escribe la pregunta tal cual.
     */
    suspend fun ask(prompt: String, images: List<File> = emptyList()): String =
        withContext(Dispatchers.IO) {
            mutex.withLock {
                val partes = mutableListOf<Content>()
                partes.add(Content.Text(prompt))
                images.forEach { img -> partes.add(Content.ImageFile(img.absolutePath)) }
                val contents = Contents.of(*partes.toTypedArray())
                engine.createConversation().use { conversation ->
                    conversation.sendMessage(contents).toString().trim()
                }
            }
        }

    /**
     * Transcribe un archivo de audio grabado con MediaRecorder, usando el
     * reconocimiento de voz nativo de Gemma 4 E2B. Sustituye por completo
     * al SpeechRecognizer del sistema operativo: no depende de qué motor
     * de voz haya elegido el fabricante del móvil (el problema detectado
     * en Samsung/One UI), porque todo el reconocimiento ocurre dentro del
     * propio modelo, 100% on-device.
     */
    suspend fun transcribe(audioFile: File, langName: String): String =
        withContext(Dispatchers.IO) {
            mutex.withLock {
                val systemInstruction = """
                    Transcribe fielmente el audio adjunto, hablado en $langName.
                    Responde ÚNICAMENTE con la transcripción exacta de lo dicho,
                    sin explicaciones, sin comillas y sin comentarios adicionales.
                """.trimIndent()
                val conversationConfig = ConversationConfig(
                    systemInstruction = Contents.of(systemInstruction),
                )
                engine.createConversation(conversationConfig).use { conversation ->
                    val contents = Contents.of(Content.AudioFile(audioFile.absolutePath))
                    conversation.sendMessage(contents).toString().trim()
                }
            }
        }

    fun close() {
        if (::engine.isInitialized) {
            engine.close()
        }
    }
}

File: ios/Sources/TranslationEngine.swift
import Foundation
import LiteRTLM

/// Envoltorio sobre el motor de LiteRT-LM para hacer traducciones puntuales
/// y transcripciones de audio, con el modelo Gemma 4 E2B. Es el equivalente
/// en Swift de TranslationEngine.kt (Android) — mismas funciones, misma idea:
/// cada llamada abre una conversación nueva para no arrastrar contexto entre
/// turnos.
///
/// Se declara como "actor" (el equivalente en Swift al Mutex que usábamos en
/// Kotlin): garantiza que solo una parte del código a la vez toca el motor,
/// aunque se llame desde varios sitios de la app al mismo tiempo.
actor TranslationEngine {
    private var engine: Engine?
    private let modelPath: String

    init(modelPath: String) {
        self.modelPath = modelPath
    }

    var isModelPresent: Bool {
        FileManager.default.fileExists(atPath: modelPath)
    }

    func initialize() async throws {
        let config = try EngineConfig(
            modelPath: modelPath,
            backend: .cpu(),
            // Necesario para poder transcribir audio (pestaña Traductor) y,
            // más adelante, procesar imágenes (pestaña Asistente IA).
            visionBackend: .cpu(),
            audioBackend: .cpu(),
            cacheDir: NSTemporaryDirectory()
        )
        let nuevoEngine = Engine(engineConfig: config)
        try await nuevoEngine.initialize()
        self.engine = nuevoEngine
    }

    /// Traduce un texto de un idioma a otro. Equivalente a translate() en
    /// TranslationEngine.kt.
    func translate(text: String, fromLangName: String, toLangName: String) async throws -> String {
        guard let engine else {
            throw NSError(
                domain: "TranslationEngine",
                code: 1,
                userInfo: [NSLocalizedDescriptionKey: "El motor no está inicializado"]
            )
        }
        let systemInstruction = """
        Eres un traductor profesional simultáneo.
        Traduce SIEMPRE del \(fromLangName) al \(toLangName).
        Responde ÚNICAMENTE con la traducción, sin explicaciones,
        sin comillas y sin repetir el texto original.
        """
        let config = ConversationConfig(systemMessage: Message(systemInstruction))
        let conversation = try await engine.createConversation(with: config)
        let response = try await conversation.sendMessage(Message(text))
        return response.toString.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// Transcribe un archivo de audio (.wav) en el idioma indicado.
    /// Equivalente a transcribe() en TranslationEngine.kt.
    func transcribe(audioPath: String, languageName: String) async throws -> String {
        guard let engine else {
            throw NSError(
                domain: "TranslationEngine",
                code: 1,
                userInfo: [NSLocalizedDescriptionKey: "El motor no está inicializado"]
            )
        }
        let prompt = "Transcribe este audio en \(languageName). Responde únicamente con la transcripción, sin explicaciones."
        let message = Message(contents: [
            Content.audioFile(audioPath),
            Content.text(prompt)
        ])
        let conversation = try await engine.createConversation()
        let response = try await conversation.sendMessage(message)
        return response.toString.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}
